### ISSUE TITLE
Error running container with own SSL and conf.yaml files (Resolved)

### ISSUE THREAD

**Issue by HaakonFlaaronning** [Voyz/ibeam#28]
**The bug**
I attempted to create and use a self-signed certificate using the steps explained in [TLS certificates and HTTPS](https://github.com/Voyz/ibeam/wiki/TLS-Certificates-and-HTTPS), and ran into some trouble when starting the docker container with the new SSL and `conf.yaml` files. The error is shown in the screenshot below.

![ibeam-err](https://user-images.githubusercontent.com/57355918/131470253-6c159a73-87ef-4c3c-b618-ed52c73b00ef.png)

**Solution**
The solution was to fetch IB's original `conf.yaml` file and use that as starting point instead of the one recommended in [Gateway Configuration](https://github.com/Voyz/ibeam/wiki/Gateway-Configuration). I changed `sslCert` and `sslPwd` properties as explained in [Certificates in Conf.yaml](https://github.com/Voyz/ibeam/wiki/TLS-Certificates-and-HTTPS#certificates-in-conf-yaml). The original IB configuration file looks like this:

```
    ip2loc: "US"
    proxyRemoteSsl: true
    proxyRemoteHost: "https://api.ibkr.com"
    listenPort: 5000
    listenSsl: true
    svcEnvironment: "v1"
    sslCert: "vertx.jks"
    sslPwd: "mywebapi"
    authDelay: 3000
    portalBaseURL: ""
    serverOptions:
        blockedThreadCheckInterval: 1000000
        eventLoopPoolSize: 20
        workerPoolSize: 20
        maxWorkerExecuteTime: 100
        internalBlockingPoolSize: 20
    cors:
        origin.allowed: "*"
        allowCredentials: false
    webApps:
        - name: "demo"
          index: "index.html"
    ips:
      allow:
        - 192.*
        - 131.216.*
        - 127.0.0.1
      deny:
        - 212.90.324.10
```

I am not sure if I am the only one experiencing this issue, but if others run into the same problem, we might want to update the recommended starting point of the `conf.yaml` file explained in [Gateway Configuration](https://github.com/Voyz/ibeam/wiki/Gateway-Configuration).

**Extra**
To make a request using the SSL certificate using Python's requests library, I had to use "localhost" as the answer to the question "What is your first and last name?" when creating the `cacert.jks` file ([Generating Certificated](https://github.com/Voyz/ibeam/wiki/TLS-Certificates-and-HTTPS#generating-certificates)). I have little prior experience with SSL so I am not sure if I should leave it like that - any input is appreciated.

Example request (working):

```
import requests

test = requests.get("https://localhost:5000/v1/api/one/user", verify='inputs/cacert.pem')
print(test.content)
```

**Comment by Voyz** [comment]
Hey @HaakonFlaaronning thanks for bringing the discussion here to GH and for contributing with this issue report 👏

I'm happy to hear you've found a solution already - well done! I'd like to understand what failed in the current `conf.yaml` recommended by IBeam. Could you share your `conf.yaml` you're using currently so I could look for discrepancies? Naturally, make sure to obscure your SSL password. Appreciate it! 🙏

----

As for your **Extra** question - I'm not sure I'm understanding what you're asking for there. Is it whether you should leave the 'first and last name' as 'localhost'? As far as I could see you should be able to ignore this question, as long as you provide correct IPs and DNSs when creating the certificate. See this part of the documentation:

> Optionally, you may want to add additional option to provide Subject Alternative Names (SAN) in order for the certificate to accept requests from your client hosts. For instance, if the server with IBeam is to be communicated with from two client machines, one with IP address of 10.148.0.0 and one with DNS of my-client.machine.com, your keytool command line should include:
> ```
>-ext SAN=ip:10.147.0.0,dns:my-client.machine.com
>```

Does this answer your question?

**Comment by HaakonFlaaronning** [comment]
Hey @Voyz! The `conf.yaml` I'm using is as follows:

```
    ip2loc: "US"
    proxyRemoteSsl: true
    proxyRemoteHost: "https://api.ibkr.com"
    listenPort: 5000
    listenSsl: true
    svcEnvironment: "v1"
    sslCert: "cacert.jks"
    sslPwd: "MY-PASSWORD"
    authDelay: 3000
    portalBaseURL: ""
    serverOptions:
        blockedThreadCheckInterval: 1000000
        eventLoopPoolSize: 20
        workerPoolSize: 20
        maxWorkerExecuteTime: 100
        internalBlockingPoolSize: 20
    cors:
        origin.allowed: "*"
        allowCredentials: false
    webApps:
        - name: "demo"
          index: "index.html"
    ips:
      allow:
        - 192.*
        - 131.216.*
        - 127.0.0.1
        - 0.0.0.0
        - 172.17.0.*
      deny:
        - 212.90.324.10
```

The error I got was due to the `ccp: false` parameter that you include in your `conf.yaml` file in the documentation.

Yes, my question was whether I should leave the "first and last name" as "localhost". If I don't I get the following error when starting the container:

``` 
urllib.error.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: Hostname mismatch, certificate is not valid for 'localhost'. (_ssl.c:1076)>
```
..even though as you can see '127.0.0.1' is included in the IP allow list.

**Comment by Voyz** [comment]
Thanks for the `conf.yaml` @HaakonFlaaronning 👍 

You're right, it's the `ccp:false` that causes this issue. I tried tracing back how did it end up in the Wiki, but I can't see any reason why it would be there. I'm imagining I've copied the beta version of the `conf.yaml` which indeed includes that line by mistake. Thanks for catching it! And sorry for the issue.

----

As for your second question - thanks for clarifying. So note that you're speaking of two different things over there:

> whether I should leave the "first and last name" as "localhost".

Here you're talking about the SSL certificate, and whether it should have `localhost` in it.

> even though as you can see '127.0.0.1' is included in the IP allow list.

Here you're talking about the `conf.yaml`, and whether it has `127.0.0.1` in it.

These are different files and are used differently. Having either `localhost` or `127.0.0.1` in one, doesn't mean it will be respected in the other place.

As far as I'm aware, you don't need to answer `localhost` to 'first and last name' question, as long as you provide `localhost` as SAN (see my previous reply), ie: generating the key with the following flag `-ext SAN=ip:127.0.0.1,dns:localhost`. Wanna give that a shot and let me know if it works?
